### PR TITLE
Ball interception v2

### PR DESCRIPTION
**Pull request by julianschuler**
## Introduced Changes

This PR improves the ball interception by calculating the shortest path for intercepting the ball while considering the robot's current movement.

## ToDo / Known Issues

- [x] Write tests for geometric functions (`calculate_arc_tangent_to` and `triangle_orientation`)
- [x] Update the scenario file for testing with the behavior simulator
- [x] Parameter tuning

## Ideas for Next Iterations (Not This PR)

None.

## How to Test

Flash a NAO or execute the `ball_intercept.lua` or `ball_intercept_striker.lua` scenario in the behavior simulator.

### CODE REVIEW

**Review by knoellle — CHANGES_REQUESTED** [review]

File: crates/types/src/geometry.rs [hunk 1]
```diff
@@ -20,6 +20,26 @@ impl Orientation {
...
+
+        if orientation == 0.0 {
+            Self::Colinear
+        } else if orientation.is_sign_negative() {
+            Self::Clockwise
+        } else {
+            Self::Counterclockwise
+        }
+    }
 }
```
[the remark this answers is not in the document]
Comment (julianschuler) [reply crates/types/src/geometry.rs]: Thanks, done.

File: crates/control/src/behavior/intercept_ball.rs [hunk 2]
```diff
@@ -69,3 +77,130 @@ pub fn execute(world_state: &WorldState, parameters: InterceptBall) -> Option<Mo
...
+
+    let ball_in_field_velocity = robot_to_field * ball.ball_in_ground_velocity;
+    let ball_moving = ball_in_field_velocity.norm() > parameters.minimum_ball_velocity;
+    let ball_moving_towards_own_half =
+        ball_in_field_velocity.x < -parameters.minimum_ball_velocity_towards_own_half;
+
+    ball_in_front_of_robot
+        && ball_moving
+        && ball_moving_towards_robot
+        && ball_moving_towards_own_half
```
[the remark this answers is not in the document]
Comment (julianschuler) [reply crates/control/src/behavior/intercept_ball.rs]: Good point, done.

File: crates/control/src/behavior/intercept_ball.rs [hunk 3]
```diff
@@ -69,3 +77,130 @@ pub fn execute(world_state: &WorldState, parameters: InterceptBall) -> Option<Mo
...
+    if arc_too_small || total_angle_change_too_lange || angle_change_too_small {
+        return vec![PathSegment::LineSegment(LineSegment(
+            Point2::origin(),
+            optimal_interception_point,
+        ))];
+    }
+
+    // If the arc and angle changes are within range, rotate current direction by angle and bridge
+    // the path between current position and line segment to interception point by a tangent arc
+    let direction = UnitComplex::new(angle_change) * walking_direction;
```
[the remark this answers is not in the document]
Comment (julianschuler) [reply crates/control/src/behavior/intercept_ball.rs]: Good point, done.

File: crates/control/src/behavior/intercept_ball.rs [hunk 4]
```diff
@@ -29,38 +35,40 @@ pub fn execute(world_state: &WorldState, parameters: InterceptBall) -> Option<Mo
...
 
+            let normalized_velocity = ball.ball_in_ground_velocity.normalize();
+
+            // Find the point with the least distance from the line traversed by the ball
+            let optimal_interception_point = ball.ball_in_ground
+                - ball.ball_in_ground.coords.dot(&normalized_velocity) * normalized_velocity;
+
+            if optimal_interception_point.coords.norm() > parameters.maximum_intercept_distance {
+                return None;
+            }
```
[the remark this answers is not in the document]
Comment (julianschuler) [reply crates/control/src/behavior/intercept_ball.rs:60]: Yes, but this requires a better ball model for the velocity decay so I would propose to do this in another iteration. We could also combine this with "time to reach ball", but this is also not fully finished yet, so I suggest to do this later.

**Review by knoellle — APPROVED** [review]
LGTM

**Pull request merged**